test/common/upstream/load_balancer_fuzz_base.cc

```diff
@@ -0,0 +1,201 @@
+  }
+
+  HostVector locality_two = {};
+  for (uint8_t index : localities[1]) {
+    ENVOY_LOG_MISC(trace, "Added host at index {} to locality 2", index);
+    locality_two.push_back(host_set.hosts_[index]);
+    locality_indexes_[index] = 1;
+  }
+
+  HostVector locality_three = {};
+  for (uint8_t index : localities[2]) {
+    ENVOY_LOG_MISC(trace, "Added host at index {} to locality 3", index);
+    locality_three.push_back(host_set.hosts_[index]);
+    locality_indexes_[index] = 2;
+  }
+
+  host_set.hosts_per_locality_ = makeHostsPerLocality({locality_one, locality_two, locality_three});
+}
+
+// Initializes random and fixed host sets
+void LoadBalancerFuzzBase::initializeLbComponents(
+    const test::common::upstream::LoadBalancerTestCase& input) {
+  random_.initializeSeed(input.seed_for_prng());
+  uint8_t priority_of_host_set = 0;
+  for (const auto& setup_priority_level : input.setup_priority_levels()) {
+    initializeASingleHostSet(setup_priority_level, priority_of_host_set);
+    priority_of_host_set++;
+  }
+  num_priority_levels_ = priority_of_host_set;
+}
+
+// Updating host sets is shared amongst all the load balancer tests. Since logically, we're just
+// setting the mock priority set to have certain values, and all load balancers interface with host
+// sets and their health statuses, this action maps to all load balancers.
+void LoadBalancerFuzzBase::updateHealthFlagsForAHostSet(const uint64_t host_priority,
+                                                        const uint32_t num_healthy_hosts,
+                                                        const uint32_t num_degraded_hosts,
+                                                        const uint32_t num_excluded_hosts,
+                                                        const std::string random_bytestring) {
+  const uint8_t priority_of_host_set = host_priority % num_priority_levels_;
+  ENVOY_LOG_MISC(trace, "Updating health flags for host set at priority: {}", priority_of_host_set);
+  MockHostSet& host_set = *priority_set_.getMockHostSet(priority_of_host_set);
+  // This downcast will not overflow because size is capped by port numbers
+  const uint32_t host_set_size = host_set.hosts_.size();
+  host_set.healthy_hosts_.clear();
+  host_set.degraded_hosts_.clear();
+  host_set.excluded_hosts_.clear();
+
+  Fuzz::ProperSubsetSelector subset_selector(random_bytestring);
+
+  const std::vector<std::vector<uint8_t>> subsets = subset_selector.constructSubsets(
+      {num_healthy_hosts, num_degraded_hosts, num_excluded_hosts}, host_set_size);
+
+  // Healthy hosts are first subset
+  for (uint8_t index : subsets.at(0)) {
+    host_set.healthy_hosts_.push_back(host_set.hosts_[index]);
+    ENVOY_LOG_MISC(trace, "Index of host made healthy at priority level {}: {}",
+                   priority_of_host_set, index);
+  }
+
+  // Degraded hosts are second subset
+  for (uint8_t index : subsets.at(1)) {
+    host_set.degraded_hosts_.push_back(host_set.hosts_[index]);
```

Comment:
are the health flags actually updated?

Reply:
After quickly glancing through the codebase, I don't think that health flags affect the load balancing algorithm. I'm not for sure on this, but it seems like Health Flags go into a lot more specificity than the logical buckets of Healthy, Degraded, and Excluded.

Reply:
huh, that's strange, because this is how those locality subsets are constructed
https://github.com/envoyproxy/envoy/blob/22683a0a24ffbb0cdeb4111eec5ec90246bec9cb/source/common/upstream/upstream_impl.cc#L956

but if they don't impact the LB, then I trust you!

Reply:
Will take a look! Sorry about being wrong!

Reply:
Yeah I understand the flags are used to bucket the hosts into healthy/degraded/excluded, but my fuzz test already does that.

Reply:
I think that this doesn't come up except in subset. https://github.com/envoyproxy/envoy/blob/22683a0a24ffbb0cdeb4111eec5ec90246bec9cb/source/common/upstream/upstream_impl.cc#L946 Here's another example. See how the flag is used to know which one of healthy/degraded/excluded as a list to add too...my fuzzer already does that with subset selection.

Reply:
Health status definitely should affect LB. Take a look at what zone aware LB does and the concept of panic routing.

Reply:
Hey, I know you two have years of experience with this codebase and wrote a good portion of it, but I feel like I'm missing something here. https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/load_balancer_impl_test.cc, https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/ring_hash_lb_test.cc, and https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/maglev_lb_test.cc have zero logic whatsoever dealing with health flags? Doesn't the lb solely interact with the healthy/degraded/excluded host lists within a priority level, otherwise the unit tests would've explicitly changed health flags? Isn't the point of the health flags to map the host to put be in one of these three HostVectors of health/degraded/excluded hosts, which is what my fuzzer does/the unit tests do explicitly?

Reply:
https://github.com/envoyproxy/envoy/blob/master/test/common/upstream/load_balancer_impl_test.cc#L418

This updates the host set by bucketing the hosts into the host vectors healthy/degraded/excluded and doesn't set the health flags at all and induces a panic across all three levels.

Reply:
Although there's only one or two instances of health flags directly being used (which i think are enough to justify it) as opposed to the pre-set host vectors of healthy/degraded/excluded, i think it's bad to rely on the fact that they will never be used. Someone could easily add an ASSERT that the host health flag of a host in those lists corresponds to healthy.

Another thing you can't do because health flags are never changed is validate that if chooseHost returns a host, the host better be healthy. You can't do that unless you do an O(n) search through the host sets unless you update their health flags as you update the sets.

Reply:
Added health flags! Thanks everyone!